### PR TITLE
Add manual limit selection to graph header/footer

### PR DESCRIPTION
**Pull request by thomasloven**
## Proposed change

Add a manual y-limit selection to graph header/footer and sensor cards.

```yaml
type: sensor
entity: sensor.temperature2
graph: line
detail: 2
hours_to_show: 1
limits:
  min: 0
  max: 30
name: 'Manual limits [0,30]'
```

Default limits still automatically fit the data unless any `unit_of_measurement` is `%`. Then the default are 0-100.

No GUI editor.

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information
Edit 2021-07-18:

A bit of background and discussion can be found on discord, starting here and continuing a few dozen messages: https://discord.com/channels/330944238910963714/351047592588869643/837837258743873577

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/lovelace/header-footer/hui-graph-header-footer.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ import {
   TemplateResult,
 } from "lit-element";
 import "../../../components/ha-circular-progress";
+import { showMediaBrowserDialog } from "../../../components/media-player/show-media-browser-dialog";
```
Comment (bramkragten) [inline src/panels/lovelace/header-footer/hui-graph-header-footer.ts:14]: And we can open the media browser now? 😜

Comment (thomasloven) [reply src/panels/lovelace/header-footer/hui-graph-header-footer.ts]: I have no idea how that got in there. Must have accidentally hit tab at the wrong moment in vscode or something...

Comment (thomasloven) [reply src/panels/lovelace/header-footer/hui-graph-header-footer.ts]: Why didn't the commit hook linting warn me about this, btw? Is it not the same as the github action?

Comment (bramkragten) [reply src/panels/lovelace/header-footer/hui-graph-header-footer.ts]: We don't do commit hook linting (yet), we just run prettier

File: src/panels/lovelace/common/graph/coordinates.ts [hunk 2]
```diff
@@ -57,15 +57,20 @@ export const coordinates = (
...
   history = history.filter((item) => !Number.isNaN(item.state));
 
-  const min = Math.min(...history.map((item) => item.state));
-  const max = Math.max(...history.map((item) => item.state));
+  const min = limits
+    ? limits.min
+    : Math.min(...history.map((item) => item.state));
+  const max = limits
+    ? limits.max
+    : Math.max(...history.map((item) => item.state));
```
[the remark this answers is not in the document]
Comment (thomasloven) [reply src/panels/lovelace/common/graph/coordinates.ts:73]: Good point.
I suppose the types needs updating too.

File: src/panels/lovelace/cards/types.ts [hunk 3]
```diff
@@ -286,6 +286,10 @@ export interface SensorCardConfig extends LovelaceCardConfig {
   detail?: number;
   theme?: string;
   hours_to_show?: number;
+  limits?: {
+    min: number;
+    max: number;
+  };
```
Comment (thomasloven) [inline src/panels/lovelace/cards/types.ts:292]:
```suggestion
  limits?: {
    min?: number;
    max?: number;
  };
```

File: src/panels/lovelace/header-footer/types.ts [hunk 4]
```diff
@@ -15,6 +15,10 @@ export interface GraphHeaderFooterConfig extends LovelaceHeaderFooterConfig {
   entity: string;
   detail?: number;
   hours_to_show?: number;
+  limits?: {
+    min: number;
+    max: number;
+  };
```
Comment (thomasloven) [inline src/panels/lovelace/header-footer/types.ts:21]:
```suggestion
  limits?: {
    min?: number;
    max?: number;
  };
```

File: src/panels/lovelace/common/graph/coordinates.ts [hunk 5]
```diff
@@ -57,15 +57,20 @@ export const coordinates = (
   history: any,
   hours: number,
   width: number,
-  detail: number
+  detail: number,
+  limits?: { min: number; max: number }
```
Comment (thomasloven) [inline src/panels/lovelace/common/graph/coordinates.ts:61]:
```suggestion
  limits?: { min?: number; max?: number }
```

**Review by balloob — APPROVED** [review]
🎉

**Pull request merged**